### PR TITLE
Assignment 7/8 - Timothy Nelson & Trevor Felchlin

### PR DESCRIPTION
**Pull request by timmytaters**
@timmytaters @tfelchlin13

### CODE REVIEW

**Comment by Eggbertx** [comment]
I will review this

**Comment by joshmerrell** [comment]
I will review this

File: Assignment/SampleData.cs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    /*Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ❌✔
+    Using LINQ, skip the first row in the People.csv. ❌✔
+    Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ❌✔*/
+    private IEnumerable<string> csvRows = File.ReadLines("People.csv").Skip(1);
+    public IEnumerable<string> CsvRows{
+        //File.ReadLines internally handles the resource management and ensures proper disposal of resources once it finishes reading the lines.
+        //It's designed to be used without requiring you to manually dispose of resources.
+        get { return csvRows; }
+        set { csvRows = value; }
+    }
```
Comment (joshmerrell) [inline Assignment/SampleData.cs:24]: Since `CsvRows` doesn't do any validation on `csvRows`, it can be merged into
```C#
public IEnumerable<string> CsvRows{ get; set; } = File.ReadLines("People.csv").Skip(1);
```

File: Assignment/SampleData.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    /*2. Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ❌✔
+    Use ISampleData.CsvRows for your data source. ❌✔
+    Don't forget the list should be unique. ❌✔
+    Sort the list alphabetically. ❌✔
+    Include a test that leverages a hardcoded list of addresses. ❌✔
+    Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ❌✔*/
+    public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()
+    {
+        IEnumerable<String> states = CsvRows.Select(CsvRows => CsvRows.Split(",")[6].Trim()).Distinct().OrderBy(state => state);
```
Comment (joshmerrell) [inline Assignment/SampleData.cs:34]: Use `string` instead of `String`. This occurs in multiple cases in SampleData.cs and in SampleDataTests.cs

**Review by joshmerrell — COMMENTED** [review]
It looks pretty good overall, with pretty good test coverage, though you're missing a test that uses LINQ to test `GetUniqueSortedListOfStatesGivenCsvRows` and an explanation comment for why you didn't use `ISampleData.CsvRows` to test `People`

1. Implement the `ISampleData.CsvRows` property, loading the data from the `People.csv` file and returning each line as a single string. ✔
   - Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
   - Using LINQ, skip the first row in the `People.csv`. ✔
   - Be sure to appropriately handle resource (`IDisposable`) items correctly if applicable. ✔

2. Implement `IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()` to return a sorted, unique list of states. ✔
   - Use `ISampleData.CsvRows` for your data source. ✔
   - Don't forget the list should be unique. ✔
   - Sort the list alphabetically. ✔
   - Include a test that leverages a hardcoded list of addresses. ✔
   - Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ❌

3. Implement `ISampleData.GetAggregateSortedListOfStatesUsingCsvRows()` to return a `string` that contains a unique, comma separated list of states. ✔
   - Use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows()` for your data source. ✔
   - Consider "selecting" only the states and calling `ToArray()` to retrieve an array of all the state names. ✔
   - Given the array, consider using `string.Join` to combine the list into a single string. ✔

4. Implement the `ISampleData.People` property to return all the items in `People.csv` as `Person` objects ✔
   - Use `ISampleData.CsvRows` as the source of the data. ✔
   - Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
   - Be sure that `Person.Address` is also populated. ✔
   - Consider using `ISampleData.CsvRows` in your test to verify your results. ❌

5. Implement `ISampleDate.FilterByEmailAddress(Predicate<string> filter)` to return a list of names where the email address matches the `filter`. ✔
   - Use `ISampleData.People` for your data source. ✔

6. Implement `ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people)` to return a `string` that contains a unique, comma-separated list of states. ✔
   - Use the `people` parameter from `ISampleData.People` property for your data source. ✔
   - At a minimum, use the `System.Linq.Enumerable.`Aggregate` LINQ method to create your result. ✔
   - Don't forget the list should be unique. ✔

7. Given the implementation of `Node` in Assignment5
   - Implement `IEnumerable<T>` to return all the items in the "circle" of items. ✔

## Fundamentals
- Ensure nullable reference types is enabled  ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers)  ✔
- Set `LangVersion` and the `TargetFramework` to the latest released versions available (preview versions optional)   ✔
- and enabled .NET analyzers for both projects ✔
- For this assignment, consider using `Assert.AreEqual<T>()` (the generic version)  ✔
- All of the above should be unit tested ✔
- Choose simplicity over complexity ❌

**Review by alyxadarr — COMMENTED** [review]
Good job overall! I think the People test can be simplified but good job hitting mostly all the marks!

Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔

Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
Using LINQ, skip the first row in the People.csv. ✔
Be sure to appropriately handle resource (IDisposable) items correctly if applicable. ✔
Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔

Use ISampleData.CsvRows for your data source. ✔
Don't forget the list should be unique. ✔
Sort the list alphabetically. ✔
Include a test that leverages a hardcoded list of addresses. ✔
Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ❌
Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔

Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
Given the array, consider using string.Join to combine the list into a single string. ✔
Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔
Use ISampleData.CsvRows as the source of the data. ✔
Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
Be sure that Person.Address is also populated. ✔
Consider using ISampleData.CsvRows in your test to verify your results. ✔
Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔

Use ISampleData.People for your data source. ✔
Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔

Use the people parameter from ISampleData.People property for your data source. ✔
At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
Don't forget the list should be unique. ✔
Given the implementation of Node in Assignment5
Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
Fundamentals
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, consider using Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ❌

File: Assignment/SampleData.cs (hunk 1 above)
Comment (Tfelchlin13) [reply Assignment/SampleData.cs:24]: Good idea thanks!

File: Assignment/SampleData.cs (hunk 2 above)
Comment (Tfelchlin13) [reply Assignment/SampleData.cs:34]: Oh thanks for catching that

File: Assignment.Tests/NodeTests.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        node.Append(2);
+        node.Append(3);
+
+        // Assert
+        Assert.Equal(2, node.Next.Value); // Check if 2 is appended after 1
+        Assert.Equal(3, node.Next.Next.Value); // Check if 3 is appended after 2
+    }
+
+    [Fact]
+    public void ChildItems_ReturnsRemainingItemsWithMaximum_MaximumItemsReturned_Params_ExpectedResult()
```
[the remark this answers is not in the document]
Comment (Tfelchlin13) [reply Assignment.Tests/NodeTests.cs:72]: good suggestion thank you

File: Assignment.Tests/SampleDataTests.cs [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Assignment.Tests;
+public class SampleDataTests
+{
+    [Fact]
+    public void SkipingFirstRow_PeopleCSV_Successful()
```
[the remark this answers is not in the document]
Comment (Tfelchlin13) [reply Assignment.Tests/SampleDataTests.cs:12]: unsure what this is for

File: Assignment/Node.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    {
+        return GetEnumerator();
+    }
+    public IEnumerable<T> ChildItems(int max)
+    {
+        int i = 0;
+        Node<T> current = this.Next; // Start from the next node (skip the current one)
+
+        while (current != this && i < max)
+        {
```
[the remark this answers is not in the document]
Comment (Tfelchlin13) [reply Assignment/Node.cs:87]: thanks

File: Assignment/Node.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    IEnumerator IEnumerable.GetEnumerator()
+    {
+        return GetEnumerator();
+    }
+    public IEnumerable<T> ChildItems(int max)
+    {
+        int i = 0;
+        Node<T> current = this.Next; // Start from the next node (skip the current one)
```
[the remark this answers is not in the document]
Comment (Tfelchlin13) [reply Assignment/Node.cs:84]: Thanks

File: Assignment/SampleData.cs [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+using System.Threading.Channels;
+using System.Text;
+
+namespace Assignment;
+
+public class SampleData : ISampleData
+{
+    // 1.Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ❌✔
+
+    /*Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ❌✔
```
[the remark this answers is not in the document]
Comment (Tfelchlin13) [reply Assignment/SampleData.cs:15]: not a bad idea

File: Assignment.Tests/SampleDataTests.cs [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+        // Mock the behavior of the CsvRows property
+        sampleData.CsvRows = csvRows;
+
+        // Act
+        var result = sampleData.GetUniqueSortedListOfStatesGivenCsvRows();
+
+        // Assert
+        var expectedStates = new List<string> { "CA", "FL", "MT" };
+        Assert.Equal(expectedStates, result);
+    }
```
[the remark this answers is not in the document]
Comment (Tfelchlin13) [reply Assignment.Tests/SampleDataTests.cs:40]: ah nice that's a great idea

File: Assignment.Tests/SampleDataTests.cs [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+        // Act
+        IEnumerable<IPerson> result = sampleData.People;
+
+        // Assert
+        IEnumerable<IPerson> expected = new List<IPerson>();
+        expected = expected.Append(new Person("Chadd", "Stennine", new Address("94148 Kings Terrace", "Long Beach", "CA", "59721"), "cstennine2@wired.com"));
+        expected = expected.Append(new Person("Karin", "Joder", new Address("03594 Florence Park", "Tampa", "FL", "71961"), "kjoder1@quantcast.com"));
+        expected = expected.Append(new Person("Priscilla", "Jenyns", new Address("7884 Corry Way", "Helena", "MT", "70577"), "pjenyns0@state.gov"));
+        Assert.Equal(expected.ElementAt(0).FirstName, result.ElementAt(0).FirstName);
+        Assert.Equal(expected.ElementAt(0).LastName, result.ElementAt(0).LastName);
```
[the remark this answers is not in the document]
Comment (Tfelchlin13) [reply Assignment.Tests/SampleDataTests.cs:87]: yea i got to look into making it shorter

File: Assignment.Tests/NodeTests.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (timmytaters) [reply Assignment.Tests/NodeTests.cs]: done

File: Assignment.Tests/SampleDataTests.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (timmytaters) [reply Assignment.Tests/SampleDataTests.cs]: I made it shorter by making an equals method for the Person class and just doing a loop :)

File: Assignment.Tests/NodeTests.cs [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    [Fact]
+    public void ChildItems_ReturnsRemainingItemsWithMaximum_MaximumItemsReturned()
+    {
+        // Arrange
+        Node<int> headNode = new(1);
+        headNode.Append(2);
+        headNode.Append(8);
+
+        // Act
+        IEnumerable<int> test = headNode.ChildItems(2);
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/NodeTests.cs:92]: There could be a good argument made for whether maximum should be inclusive or exclusive.

File: Assignment.Tests/NodeTests.cs [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        node.Append(2);
+        node.Append(3);
+
+        // Assert
+        Assert.Equal(2, node.Next.Value); // Check if 2 is appended after 1
+        Assert.Equal(3, node.Next.Next.Value); // Check if 3 is appended after 2
+    }
+
+    [Fact]
+    public void ChildItems_ReturnsRemainingItemsWithMaximum_MaximumItemsReturned()
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/NodeTests.cs:84]: What about testing edge conditions for ChildItems? And what about making sure the items returned are what you expect?

File: Assignment.Tests/Assignment.Tests.csproj [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+	<PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
+		<NoWarn>1701;1702;CA1707</NoWarn>
+		<WarningsNotAsErrors>CA1707</WarningsNotAsErrors>
+	</PropertyGroup>
+
+	<PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|AnyCPU'">
+		<NoWarn>1701;1702;CA1707</NoWarn>
+		<WarningsNotAsErrors>CA1707</WarningsNotAsErrors>
+	</PropertyGroup>
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/Assignment.Tests.csproj:35]: Where is the justification for ignoring these warnings?

File: Assignment/SampleData.cs [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+using System.Threading.Channels;
+using System.Text;
+
+namespace Assignment;
+
+public class SampleData : ISampleData
+{
+    //File.ReadLines internally handles the resource management and ensures proper disposal of resources once it finishes reading the lines.
+    //It's designed to be used without requiring you to manually dispose of resources.
+    public IEnumerable<string> CsvRows { get; set; } = File.ReadLines("People.csv").Skip(1);
```
Comment (BenjaminMichaelis) [inline Assignment/SampleData.cs:15]: This having a public set seems a bit odd... I think at least a private set through a constructor would be better imo so data can't be modified on the fly after the object is instantiated

File: Assignment/SampleData.cs [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+
+    public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()
+    {
+        IEnumerable<string> states = CsvRows.Select(CsvRows => CsvRows.Split(",")[6].Trim()).Distinct().OrderBy(state => state);
+        return states;
```
Comment (BenjaminMichaelis) [inline Assignment/SampleData.cs:20]:
```suggestion
        return CsvRows.Select(CsvRows => CsvRows.Split(",")[6].Trim()).Distinct().OrderBy(state => state);
```
nit: simplify

Even simpler, a expression-bodied member

File: Assignment/SampleData.cs [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+
+public class SampleData : ISampleData
+{
+    //File.ReadLines internally handles the resource management and ensures proper disposal of resources once it finishes reading the lines.
+    //It's designed to be used without requiring you to manually dispose of resources.
+    public IEnumerable<string> CsvRows { get; set; } = File.ReadLines("People.csv").Skip(1);
+
+    public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()
+    {
+        IEnumerable<string> states = CsvRows.Select(CsvRows => CsvRows.Split(",")[6].Trim()).Distinct().OrderBy(state => state);
```
Comment (BenjaminMichaelis) [inline Assignment/SampleData.cs:19]: for specific columns in this file, a class or enum would be nice to give readable values instead of like '6' or '2' to different columns, so you could say like `[Column.State]` or `[Column.City]`, etc.

File: Assignment/SampleData.cs [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+        {
+            IEnumerable<IPerson> peopleOut = new List<IPerson>();
+            IEnumerable<string[]> peopleIn = CsvRows.Select(CsvRows => CsvRows.Split(','));
+            foreach(string[] person in peopleIn) 
+            { 
+                string[] curPerson = person;
+                peopleOut = peopleOut.Append(new Person(curPerson[1], curPerson[2], new Address(curPerson[4], curPerson[5], curPerson[6], curPerson[7]), curPerson[3]));
+            }
+            peopleOut = peopleOut.OrderBy(p => p.Address.State).ThenBy(p => p.Address.City).ThenBy(p => p.Address.Zip);
+            return peopleOut;
```
Comment (BenjaminMichaelis) [inline Assignment/SampleData.cs:41]:
```suggestion
            return peopleOut.OrderBy(p => p.Address.State).ThenBy(p => p.Address.City).ThenBy(p => p.Address.Zip);
```
nit: Simplify

File: Assignment/Person.cs [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+        LastName = lastName;
+        Address = address;
+        EmailAddress = emailAddress;
+    }
+
+    public string FirstName { get; set; }
+    public string LastName { get; set; }
+    public IAddress Address { get;set; }
+    public string EmailAddress { get; set; }
+    public bool ManualIsEqual(IPerson other)
```
Comment (BenjaminMichaelis) [inline Assignment/Person.cs:20]: This is strange... Why not just implement like IComparable<>? Or if it needs identity, make it a record class
